### PR TITLE
[SPARK-17513] [STREAMING] [SQL] Make StreamExecution garbage-collect its metadata

### PR DESCRIPTION
**Pull request by frreiss**
## What changes were proposed in this pull request?

This PR modifies StreamExecution such that it discards metadata for batches that have already been fully processed. I used the `purge` method that was added as part of SPARK-17235.
## How was this patch tested?

I added a test case to verify that old metadata log files are correctly purged. 
I also ran the entire regression suite.

### CODE REVIEW

**Comment by rxin** [comment]
Jenkins, test this please.

**Comment by rxin** [comment]
Looks pretty good to me.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala [hunk 1]
```diff
@@ -125,6 +125,32 @@ class StreamingQuerySuite extends StreamTest with BeforeAndAfter {
...
+      AddData(inputData, 4, 6),
+      CheckAnswer(6, 3, 6, 3, 1, 1),
+
+      // Three batches have run, but only one set of metadata should be present
+      AssertOnQuery(
+        q => {
+          val metadataLogDir = new java.io.File(q.offsetLog.metadataPath.toString)
+          val logFileNames = metadataLogDir.listFiles().toSeq.map(_.getName())
+          val toTest = logFileNames.filter(! _.endsWith(".crc")) // Workaround for SPARK-17475
+          toTest.size == 1 && toTest.head == "2"
```
Comment (petermaxlee) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala]: actually this test case will always pass, even without the change here. I will submit a patch based on yours that fixes the issues here.

**Review by petermaxlee — CHANGES_REQUESTED** [review]

**Comment by rxin** [comment]
@frreiss can you close this now?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala (hunk 1 above)
Comment (frreiss) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala]: Oops, didn't notice I had left that "true" at the end of that block of code. @petermaxlee , let me know if you need help preparing the final version for merge.

**Pull request closed without merging**